bug: data race between received mouse events and using cursed renderer

**Describe the bug**
It looks as though there is a data race when mouse events are received, when using the cursed renderer (& potentially others). `cursedRenderer` stores the last rendered `View` in `lastView` so `onMouse` can forward `MouseMsg` to `View.OnMouse`. That field is written inside `flush`, which correctly holds `s.mu` from start to finish -- including the assignment `s.lastView = &view`. `onMouse`, however, reads `s.lastView` (and dereferences it) **without** acquiring `s.mu`.

**Setup**
Please complete the following information along with version numbers, if applicable.
 - OS: n/a
 - Shell: n/a
 - Terminal Emulator: n/a
 - Terminal Multiplexer: n/a

**To Reproduce**
This test reproduces:

```go
package tea

import (
	"fmt"
	"io"
	"testing"
	"time"
)

// mouseRaceModel drives the real Bubble Tea renderer: every Update changes View
// content so flush rewrites lastView; MouseMode is set so mouse handling stays
// wired through cursedRenderer.onMouse (see go switch on MouseMsg).
type mouseRaceModel struct {
	i int
}

func (m *mouseRaceModel) Init() Cmd { return nil }

func (m *mouseRaceModel) Update(msg Msg) (Model, Cmd) {
	switch msg.(type) {
	case MouseClickMsg, MouseMotionMsg, MouseWheelMsg:
		m.i++
	}
	return m, nil
}

func (m *mouseRaceModel) View() View {
	return View{
		Content:   fmt.Sprintf("tick-%d\n", m.i),
		MouseMode: MouseModeCellMotion,
	}
}

func TestBubbleTea_cursedRenderer_mouseVsFlush(t *testing.T) {
	t.Parallel()

	// Block the input read loop without feeding bytes (same idea as a PTY that
	// has not yet sent data). EOF would tear the program down early.
	pr, pw := io.Pipe()
	defer func() { _ = pw.Close() }()

	m := &mouseRaceModel{}
	p := NewProgram(
		m,
		WithContext(t.Context()),
		WithInput(pr),
		WithOutput(io.Discard),
		WithEnvironment([]string{
			"TERM=xterm-256color",
			"TERM_PROGRAM=Apple_Terminal",
		}),
		WithoutSignals(),
		WithWindowSize(80, 24),
	)

	runDone := make(chan struct{})
	go func() {
		defer close(runDone)
		_, _ = p.Run()
	}()

	// Wait for ticker + first frames (renderer started in Run).
	time.Sleep(150 * time.Millisecond)

	const iterations = 100
	for i := range iterations {
		switch i % 4 {
		case 0:
			p.Send(MouseClickMsg{X: i % 80, Y: i % 24, Button: MouseLeft})
		case 1:
			p.Send(MouseMotionMsg{X: i % 80, Y: i % 24})
		case 2:
			p.Send(MouseWheelMsg{X: 0, Y: 0, Button: MouseWheelUp})
		default:
			p.Send(MouseReleaseMsg{X: i % 80, Y: i % 24, Button: MouseLeft})
		}
	}

	p.Quit()
	select {
	case <-runDone:
	case <-time.After(5 * time.Second):
		t.Fatal("program did not exit after Quit")
	}
}
```

**Expected behavior**
No data races 😄 

**Additional context**
<details>
<summary>go test -race output</summary>

May need to change `-count` as it's not always going to be 100% of the time.

```
$ go test -race -count 1 -run TestBubbleTea_cursedRenderer_mouseVsFlush ./...
==================
WARNING: DATA RACE
Read at 0x00c000288050 by goroutine 10:
  charm.land/bubbletea/v2.(*cursedRenderer).onMouse()
      fork-bubbletea/cursed_renderer.go:767 +0x44
  charm.land/bubbletea/v2.(*Program).eventLoop()
      fork-bubbletea/tea.go:805 +0x1727
  charm.land/bubbletea/v2.(*Program).Run()
      fork-bubbletea/tea.go:1144 +0x1479
  charm.land/bubbletea/v2.TestBubbleTea_cursedRenderer_mouseVsFlush.func2()
      fork-bubbletea/datarace_test.go:63 +0x7b

Previous write at 0x00c000288050 by goroutine 18:
  charm.land/bubbletea/v2.(*cursedRenderer).flush()
      fork-bubbletea/cursed_renderer.go:573 +0x30af
  github.com/charmbracelet/ultraviolet.printString[go.shape.string]()
      /home/liam/go/pkg/mod/github.com/charmbracelet/ultraviolet@v0.0.0-20260416155717-489999b90468/styled.go:164 +0xa0a
  [...TRUNCATED...]
  github.com/charmbracelet/ultraviolet.printString[go.shape.string]()
      /home/liam/go/pkg/mod/github.com/charmbracelet/ultraviolet@v0.0.0-20260416155717-489999b90468/styled.go:164 +0xa0a
  github.com/charmbracelet/ultraviolet.(*StyledString).Draw()
      /home/liam/go/pkg/mod/github.com/charmbracelet/ultraviolet@v0.0.0-20260416155717-489999b90468/styled.go:61 +0x265
  [...TRUNCATED...]
  github.com/charmbracelet/ultraviolet.(*StyledString).Draw()
      /home/liam/go/pkg/mod/github.com/charmbracelet/ultraviolet@v0.0.0-20260416155717-489999b90468/styled.go:54 +0xc4
  charm.land/bubbletea/v2.(*cursedRenderer).flush()
      fork-bubbletea/cursed_renderer.go:311 +0x890
  charm.land/bubbletea/v2.(*Program).startRenderer.func1()
      fork-bubbletea/tea.go:1418 +0x66

Goroutine 10 (running) created at:
  charm.land/bubbletea/v2.TestBubbleTea_cursedRenderer_mouseVsFlush()
      fork-bubbletea/datarace_test.go:61 +0x69c
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:2036 +0x21c
  testing.(*T).Run.gowrap1()
      /usr/local/go/src/testing/testing.go:2101 +0x38

Goroutine 18 (running) created at:
  charm.land/bubbletea/v2.(*Program).startRenderer()
      fork-bubbletea/tea.go:1409 +0x1ab
  charm.land/bubbletea/v2.(*Program).Run()
      fork-bubbletea/tea.go:1107 +0x11d5
  charm.land/bubbletea/v2.TestBubbleTea_cursedRenderer_mouseVsFlush.func2()
      fork-bubbletea/datarace_test.go:63 +0x7b
==================
==================
WARNING: DATA RACE
Read at 0x00c0002fc2b0 by goroutine 10:
  charm.land/bubbletea/v2.(*cursedRenderer).onMouse()
      fork-bubbletea/cursed_renderer.go:767 +0x73
  charm.land/bubbletea/v2.(*Program).eventLoop()
      fork-bubbletea/tea.go:805 +0x1727
  charm.land/bubbletea/v2.(*Program).Run()
      fork-bubbletea/tea.go:1144 +0x1479
  charm.land/bubbletea/v2.TestBubbleTea_cursedRenderer_mouseVsFlush.func2()
      fork-bubbletea/datarace_test.go:63 +0x7b

Previous write at 0x00c0002fc2b0 by goroutine 18:
  charm.land/bubbletea/v2.(*cursedRenderer).flush()
      fork-bubbletea/cursed_renderer.go:261 +0x104
  charm.land/bubbletea/v2.(*Program).startRenderer.func1()
      fork-bubbletea/tea.go:1418 +0x66

Goroutine 10 (running) created at:
  charm.land/bubbletea/v2.TestBubbleTea_cursedRenderer_mouseVsFlush()
      fork-bubbletea/datarace_test.go:61 +0x69c
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:2036 +0x21c
  testing.(*T).Run.gowrap1()
      /usr/local/go/src/testing/testing.go:2101 +0x38

Goroutine 18 (running) created at:
  charm.land/bubbletea/v2.(*Program).startRenderer()
      fork-bubbletea/tea.go:1409 +0x1ab
  charm.land/bubbletea/v2.(*Program).Run()
      fork-bubbletea/tea.go:1107 +0x11d5
  charm.land/bubbletea/v2.TestBubbleTea_cursedRenderer_mouseVsFlush.func2()
      fork-bubbletea/datarace_test.go:63 +0x7b
==================
--- FAIL: TestBubbleTea_cursedRenderer_mouseVsFlush (0.15s)
    testing.go:1712: race detected during execution of test
```

</details>

--- a/Taskfile.yaml
+++ b/Taskfile.yaml
@@ -1,6 +1,6 @@
 # https://taskfile.dev
 
-version: '3'
+version: "3"
 
 tasks:
   lint:
@@ -11,4 +11,4 @@ tasks:
   test:
     desc: Run tests
     cmds:
-      - go test ./... {{.CLI_ARGS}}
+      - go test -race -count 4 -cpu 1,4 ./... {{.CLI_ARGS}}

--- a/commands_test.go
+++ b/commands_test.go
@@ -6,6 +6,7 @@ import (
 )
 
 func TestEvery(t *testing.T) {
+	t.Parallel()
 	expected := "every ms"
 	msg := Every(time.Millisecond, func(t time.Time) Msg {
 		return expected
@@ -16,6 +17,7 @@ func TestEvery(t *testing.T) {
 }
 
 func TestTick(t *testing.T) {
+	t.Parallel()
 	expected := "tick"
 	msg := Tick(time.Millisecond, func(t time.Time) Msg {
 		return expected
@@ -26,31 +28,37 @@ func TestTick(t *testing.T) {
 }
 
 func TestBatch(t *testing.T) {
+	t.Parallel()
 	testMultipleCommands[BatchMsg](t, Batch)
 }
 
 func TestSequence(t *testing.T) {
+	t.Parallel()
 	testMultipleCommands[sequenceMsg](t, Sequence)
 }
 
 func testMultipleCommands[T ~[]Cmd](t *testing.T, createFn func(cmd ...Cmd) Cmd) {
 	t.Run("nil cmd", func(t *testing.T) {
+		t.Parallel()
 		if b := createFn(nil); b != nil {
 			t.Fatalf("expected nil, got %+v", b)
 		}
 	})
 	t.Run("empty cmd", func(t *testing.T) {
+		t.Parallel()
 		if b := createFn(); b != nil {
 			t.Fatalf("expected nil, got %+v", b)
 		}
 	})
 	t.Run("single cmd", func(t *testing.T) {
+		t.Parallel()
 		b := createFn(Quit)()
 		if _, ok := b.(QuitMsg); !ok {
 			t.Fatalf("expected a QuitMsg, got %T", b)
 		}
 	})
 	t.Run("mixed nil cmds", func(t *testing.T) {
+		t.Parallel()
 		b := createFn(nil, Quit, nil, Quit, nil, nil)()
 		if l := len(b.(T)); l != 2 {
 			t.Fatalf("expected a []Cmd with len 2, got %d", l)

--- a/cursed_renderer.go
+++ b/cursed_renderer.go
@@ -764,8 +764,14 @@ func (s *cursedRenderer) insertAbove(str string) error {
 
 // onMouse implements renderer.
 func (s *cursedRenderer) onMouse(m MouseMsg) Cmd {
-	if s.lastView != nil && s.lastView.OnMouse != nil {
-		return s.lastView.OnMouse(m)
+	var onMouse func(MouseMsg) Cmd
+	s.mu.Lock()
+	if s.lastView != nil {
+		onMouse = s.lastView.OnMouse
+	}
+	s.mu.Unlock()
+	if onMouse != nil {
+		return onMouse(m)
 	}
 	return nil
 }

--- a/cursed_renderer_test.go
+++ b/cursed_renderer_test.go
@@ -1,0 +1,80 @@
+package tea
+
+import (
+	"fmt"
+	"io"
+	"testing"
+	"time"
+)
+
+type mouseRaceModel struct {
+	i int
+}
+
+func (m *mouseRaceModel) Init() Cmd { return nil }
+
+func (m *mouseRaceModel) Update(msg Msg) (Model, Cmd) {
+	switch msg.(type) {
+	case MouseClickMsg, MouseMotionMsg, MouseWheelMsg:
+		m.i++
+	}
+	return m, nil
+}
+
+func (m *mouseRaceModel) View() View {
+	return View{
+		Content:   fmt.Sprintf("tick-%d\n", m.i),
+		MouseMode: MouseModeCellMotion,
+	}
+}
+
+// Fixes: https://github.com/charmbracelet/bubbletea/issues/1690
+func TestCursedRenderer_mouseVsFlush(t *testing.T) {
+	t.Parallel()
+
+	pr, pw := io.Pipe()
+	defer func() { _ = pw.Close() }()
+
+	m := &mouseRaceModel{}
+	p := NewProgram(
+		m,
+		WithContext(t.Context()),
+		WithInput(pr),
+		WithOutput(io.Discard),
+		WithEnvironment([]string{
+			"TERM=xterm-256color",
+			"TERM_PROGRAM=Apple_Terminal",
+		}),
+		WithoutSignals(),
+		WithWindowSize(80, 24),
+	)
+
+	runDone := make(chan struct{})
+	go func() {
+		defer close(runDone)
+		_, _ = p.Run()
+	}()
+
+	time.Sleep(150 * time.Millisecond)
+
+	const iterations = 100
+	for i := range iterations {
+		switch i % 4 {
+		case 0:
+			p.Send(MouseClickMsg{X: i % 80, Y: i % 24, Button: MouseLeft})
+		case 1:
+			p.Send(MouseMotionMsg{X: i % 80, Y: i % 24})
+		case 2:
+			p.Send(MouseWheelMsg{X: 0, Y: 0, Button: MouseWheelUp})
+		default:
+			p.Send(MouseReleaseMsg{X: i % 80, Y: i % 24, Button: MouseLeft})
+		}
+	}
+
+	p.Quit()
+	select {
+	case <-runDone:
+	case <-time.After(5 * time.Second):
+		t.Fatal("program did not exit after Quit")
+	}
+}

--- a/examples/go.mod
+++ b/examples/go.mod
@@ -23,7 +23,7 @@ require (
 	github.com/alecthomas/chroma/v2 v2.14.0 // indirect
 	github.com/atotto/clipboard v0.1.4 // indirect
 	github.com/aymerick/douceur v0.2.0 // indirect
-	github.com/charmbracelet/ultraviolet v0.0.0-20260413133134-73592393e1ad // indirect
+	github.com/charmbracelet/ultraviolet v0.0.0-20260416155717-489999b90468 // indirect
 	github.com/charmbracelet/x/exp/slice v0.0.0-20250327172914-2fdc97757edf // indirect
 	github.com/charmbracelet/x/termios v0.1.1 // indirect
 	github.com/charmbracelet/x/windows v0.2.2 // indirect

--- a/exec_test.go
+++ b/exec_test.go
@@ -90,6 +90,7 @@ func TestTeaExec(t *testing.T) {
 
 	for _, test := range tests {
 		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
 			var buf bytes.Buffer
 			var in bytes.Buffer
 
@@ -118,6 +119,7 @@ func TestTeaExec(t *testing.T) {
 }
 
 func TestTeaExecWithNilInput(t *testing.T) {
+	t.Parallel()
 	var buf bytes.Buffer
 
 	m := &testExecNoInputModel{}

--- a/options_test.go
+++ b/options_test.go
@@ -10,6 +10,7 @@ import (
 
 func TestOptions(t *testing.T) {
 	t.Run("output", func(t *testing.T) {
+		t.Parallel()
 		var b bytes.Buffer
 		p := NewProgram(nil, WithOutput(&b))
 		if f, ok := p.output.(*os.File); ok {
@@ -18,6 +19,7 @@ func TestOptions(t *testing.T) {
 	})
 
 	t.Run("renderer", func(t *testing.T) {
+		t.Parallel()
 		p := NewProgram(nil, WithoutRenderer())
 		if !p.disableRenderer {
 			t.Errorf("expected renderer to be a nilRenderer, got %v", p.renderer)
@@ -25,6 +27,7 @@ func TestOptions(t *testing.T) {
 	})
 
 	t.Run("without signals", func(t *testing.T) {
+		t.Parallel()
 		p := NewProgram(nil, WithoutSignals())
 		if atomic.LoadUint32(&p.ignoreSignals) == 0 {
 			t.Errorf("ignore signals should have been set")
@@ -32,6 +35,7 @@ func TestOptions(t *testing.T) {
 	})
 
 	t.Run("filter", func(t *testing.T) {
+		t.Parallel()
 		p := NewProgram(nil, WithFilter(func(_ Model, msg Msg) Msg { return msg }))
 		if p.filter == nil {
 			t.Errorf("expected filter to be set")
@@ -39,6 +43,7 @@ func TestOptions(t *testing.T) {
 	})
 
 	t.Run("external context", func(t *testing.T) {
+		t.Parallel()
 		extCtx, extCancel := context.WithCancel(context.Background())
 		defer extCancel()
 
@@ -55,6 +60,7 @@ func TestOptions(t *testing.T) {
 		}
 
 		t.Run("nil input", func(t *testing.T) {
+			t.Parallel()
 			exercise(t, WithInput(nil), func(p *Program) {
 				if !p.disableInput || p.input != nil {
 					t.Errorf("expected input to be disabled, got %v", p.input)
@@ -63,6 +69,7 @@ func TestOptions(t *testing.T) {
 		})
 
 		t.Run("custom input", func(t *testing.T) {
+			t.Parallel()
 			var b bytes.Buffer
 			exercise(t, WithInput(&b), func(p *Program) {
 				if p.input != &b {
@@ -79,6 +86,7 @@ func TestOptions(t *testing.T) {
 		}
 
 		t.Run("without catch panics", func(t *testing.T) {
+			t.Parallel()
 			exercise(t, WithoutCatchPanics(), func(p *Program) {
 				if !p.disableCatchPanics {
 					t.Errorf("expected catch panics to be disabled")
@@ -87,6 +95,7 @@ func TestOptions(t *testing.T) {
 		})
 
 		t.Run("without signal handler", func(t *testing.T) {
+			t.Parallel()
 			exercise(t, WithoutSignalHandler(), func(p *Program) {
 				if !p.disableSignalHandler {
 					t.Errorf("expected signal handler to be disabled")

--- a/screen_test.go
+++ b/screen_test.go
@@ -132,6 +132,7 @@ func TestViewModel(t *testing.T) {
 
 	for _, test := range tests {
 		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
 			var buf bytes.Buffer
 			var in bytes.Buffer
 
@@ -179,6 +180,7 @@ func TestClearMsg(t *testing.T) {
 
 	for _, test := range tests {
 		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
 			var buf bytes.Buffer
 			var in bytes.Buffer
 

--- a/tea_test.go
+++ b/tea_test.go
@@ -4,6 +4,7 @@ import (
 	"bytes"
 	"context"
 	"errors"
+	"fmt"
 	"io"
 	"strings"
 	"sync"
@@ -66,6 +67,7 @@ func (m *testModel) View() View {
 }
 
 func TestTeaModel(t *testing.T) {
+	t.Parallel()
 	var buf bytes.Buffer
 	var in bytes.Buffer
 	in.Write([]byte("q"))
@@ -88,6 +90,7 @@ func TestTeaModel(t *testing.T) {
 }
 
 func TestTeaQuit(t *testing.T) {
+	t.Parallel()
 	var buf bytes.Buffer
 	var in bytes.Buffer
 
@@ -112,6 +115,7 @@ func TestTeaQuit(t *testing.T) {
 }
 
 func TestTeaWaitQuit(t *testing.T) {
+	t.Parallel()
 	var buf bytes.Buffer
 	var in bytes.Buffer
 
@@ -165,6 +169,7 @@ func TestTeaWaitQuit(t *testing.T) {
 }
 
 func TestTeaWaitKill(t *testing.T) {
+	t.Parallel()
 	var buf bytes.Buffer
 	var in bytes.Buffer
 
@@ -218,9 +223,12 @@ func TestTeaWaitKill(t *testing.T) {
 }
 
 func TestTeaWithFilter(t *testing.T) {
-	testTeaWithFilter(t, 0)
-	testTeaWithFilter(t, 1)
-	testTeaWithFilter(t, 2)
+	for _, preventCount := range []uint32{0, 1, 2} {
+		t.Run(fmt.Sprintf("prevent_%d", preventCount), func(t *testing.T) {
+			t.Parallel()
+			testTeaWithFilter(t, preventCount)
+		})
+	}
 }
 
 func testTeaWithFilter(t *testing.T, preventCount uint32) {
@@ -260,6 +268,7 @@ func testTeaWithFilter(t *testing.T, preventCount uint32) {
 }
 
 func TestTeaKill(t *testing.T) {
+	t.Parallel()
 	var buf bytes.Buffer
 	var in bytes.Buffer
 
@@ -292,6 +301,7 @@ func TestTeaKill(t *testing.T) {
 }
 
 func TestTeaContext(t *testing.T) {
+	t.Parallel()
 	ctx, cancel := context.WithCancel(t.Context())
 	var buf bytes.Buffer
 	var in bytes.Buffer
@@ -325,6 +335,7 @@ func TestTeaContext(t *testing.T) {
 }
 
 func TestTeaContextImplodeDeadlock(t *testing.T) {
+	t.Parallel()
 	ctx, cancel := context.WithCancel(t.Context())
 	var buf bytes.Buffer
 	var in bytes.Buffer
@@ -351,6 +362,7 @@ func TestTeaContextImplodeDeadlock(t *testing.T) {
 }
 
 func TestTeaContextBatchDeadlock(t *testing.T) {
+	t.Parallel()
 	ctx, cancel := context.WithCancel(t.Context())
 	var buf bytes.Buffer
 	var in bytes.Buffer
@@ -386,6 +398,7 @@ func TestTeaContextBatchDeadlock(t *testing.T) {
 }
 
 func TestTeaBatchMsg(t *testing.T) {
+	t.Parallel()
 	var buf bytes.Buffer
 	var in bytes.Buffer
 
@@ -421,6 +434,7 @@ func TestTeaBatchMsg(t *testing.T) {
 }
 
 func TestTeaSequenceMsg(t *testing.T) {
+	t.Parallel()
 	var buf bytes.Buffer
 	var in bytes.Buffer
 
@@ -445,6 +459,7 @@ func TestTeaSequenceMsg(t *testing.T) {
 }
 
 func TestTeaSequenceMsgWithBatchMsg(t *testing.T) {
+	t.Parallel()
 	var buf bytes.Buffer
 	var in bytes.Buffer
 
@@ -472,6 +487,7 @@ func TestTeaSequenceMsgWithBatchMsg(t *testing.T) {
 }
 
 func TestTeaNestedSequenceMsg(t *testing.T) {
+	t.Parallel()
 	var buf bytes.Buffer
 	var in bytes.Buffer
 
@@ -496,6 +512,7 @@ func TestTeaNestedSequenceMsg(t *testing.T) {
 }
 
 func TestTeaSend(t *testing.T) {
+	t.Parallel()
 	var buf bytes.Buffer
 	var in bytes.Buffer
 
@@ -517,6 +534,7 @@ func TestTeaSend(t *testing.T) {
 }
 
 func TestTeaNoRun(t *testing.T) {
+	t.Parallel()
 	var buf bytes.Buffer
 	var in bytes.Buffer
 
@@ -528,6 +546,7 @@ func TestTeaNoRun(t *testing.T) {
 }
 
 func TestTeaPanic(t *testing.T) {
+	t.Parallel()
 	var buf bytes.Buffer
 	var in bytes.Buffer
 
@@ -558,6 +577,7 @@ func TestTeaPanic(t *testing.T) {
 }
 
 func TestTeaGoroutinePanic(t *testing.T) {
+	t.Parallel()
 	var buf bytes.Buffer
 	var in bytes.Buffer
 

